lib/bootloader_setup.pm

```diff
@@ -56,6 +58,91 @@ sub stop_grub_timeout {
     send_key 'up';
 }
 
+=head2 add_custom_grub_entries
+
+Add custom grub entries with extra kernel parameters.
+It adds 3rd line with default options + 4th line with advanced options.
+Extra kernel parameters are taken in GRUB_PARAM variable.
+
+e.g.  grub entries before:
+* SLES 15
+* Advanced options for SLES 15
+* Start bootloader from a read-only snapshot
+
+grub entries with GRUB_PARAM='ima_policy=tcb' and calling add_custom_grub_entries:
+* SLES 15
+* Advanced options for SLES 15
+* SLES 15 (ima_policy=tcb)
+* Advanced options for SLES 15 (ima_policy=tcb)
+* Start bootloader from a read-only snapshot
+
+And of course the new entries have ima_policy=tcb added to kernel parameters.
+
+=cut
+sub add_custom_grub_entries {
+    my $grub_param = get_var('GRUB_PARAM');
+    return unless defined($grub_param);
+
+    assert_script_run('cp /boot/grub2/grub.cfg grub.cfg.old');
+    upload_logs('grub.cfg.old', failok => 1);
+
+    assert_script_run('cp /etc/grub.d/40_custom 40_custom.tmp');
+    assert_script_run('cp /etc/grub.d/10_linux /etc/grub.d/11_linux_openqa');
+
+    my $cmd = "sed -i -e 's/\\(args=.\\)\\(\\\$4\\)/\\1$grub_param \\2/'";
+    $cmd .= " -e 's/\\(Advanced options for %s\\)/\\1 ($grub_param)/'";
+    $cmd .= " -e 's/\\(menuentry .\\\$(echo .\\\$title\\)/\\1 ($grub_param)/'";
+    $cmd .= " -e 's/\\(menuentry .\\\$(echo .\\\$os\\)/\\1 ($grub_param)/' /etc/grub.d/11_linux_openqa";
+    assert_script_run($cmd);
+    upload_logs('/etc/grub.d/11_linux_openqa', failok => 1);
+    assert_script_run('grub2-mkconfig -o /boot/grub2/grub.cfg');
+    upload_logs('/boot/grub2/grub.cfg', failok => 1);
+
+    my $distro = (is_sle() ? "SLES" : "openSUSE");
+    $distro .= " " if !is_pre_15();
+    $distro .= get_required_var('VERSION');
+    my $cnt_orig = script_output("grep -c 'menuentry .$distro' grub.cfg.old");
+    my $cnt_new  = script_output("grep -c 'menuentry .$distro' /boot/grub2/grub.cfg");
+    die("Unexpected number of grub entries: $cnt_new, expected: " . ($cnt_orig * 2)) if ($cnt_orig * 2 != $cnt_new);
```

Comment:
You should not rely on `script_output` returning exactly one token. It might include additional output on the serial port which you did not expect. Looking for a regex in the returned output is normally fine but taking the whole output as is and comparing it will not work as you expect

Reply:
I did quite lot of testing on different distros and it works well. If this happen, shouldn't it be fixed in `script_output`? I'd really prefer to keep it until any error proven. This can be saved by regexp (read only `^\d+` and ignore the rest or having optional parameter to avoid checks).

Reply:
https://progress.opensuse.org/issues/30613 is the corresponding ticket for "can it not be fixed in script_output" and it's 5 months old and so far the answer is: "We don't know how". Seriously, trust me here: You don't want this to fail. But of course you are free to test more on ppc64le and aarch64 where we saw these problems more often :)